Inference update post (#646)
huggingface/blog: inference-update.md

        Update on GitHub
    </a>
</div>

<div class="author-card">
        <a href="https://twitter.com/julsimon">
        <img class="avatar avatar-user" src="https://aeiljuispo.cloudimg.io/v7/https://s3.amazonaws.com/moonup/production/uploads/1633343465505-noauth.jpeg?w=128&h=128&f=face" title="Julien Simon">
        <div class="bfc">
            <code>juliensimon</code>
            <span class=fullname">Julien Simon</span>
        </div>
    </a>
</div>

Transformer models are quickly becoming the de facto architecture for a wide range of machine learning (ML) applications, including natural language processing, computer vision, speech, and more. Every day, developers and organizations are adopting these models to turn ideas into proof-of-concept demos, and demos into production-grade applications. 

At Hugging Face, we are obsessed with simplifying ML development and operations, without compromising on state-of-the-art performance. In this respect, testing and deploying models with minimal friction is critical all along the lifecycle of your ML project. 

Let's look at your options.

> ## Inference Widget

Reviewer: ```suggestion
## Free Inference with Model Card Widgets
```

Reviewer: or "Free Inference Widget"

Author: Free Inference Widget

One of my favorite features on the Hugging Face hub is the Inference [Widget](https://huggingface.co/docs/hub/models-widgets). Located on the model page, the Inference Widget lets you upload sample data and predict it in a single click. It's the best way to quickly get a sense of what a model does, what its output looks like, and how it performs on a few samples from your dataset. The model is loaded on-demand on our servers and unloaded when it's not needed anymore. You don't have to write any code and the feature is free. What's not to love?

> Today, we're really happy to announce that our friends at [Intel](https://www.intel.com/content/www/us/en/homepage.html) are now sponsoring inference for all CPU-based models. 

Reviewer: ```suggestion
Today, we're really happy to announce that our friends at [Intel](https://huggingface.co/intel) are now sponsoring inference for all CPU-based models. 
```

> Today, we're really happy to announce that our friends at [Intel](https://www.intel.com/content/www/us/en/homepage.html) are now sponsoring inference for all CPU-based models. 

Reviewer: I would elevate that message before "here are our options", not just to make Intel happy but also because they are sponsoring both the widget and API experiences (on CPU), and we can message that free accessibility is important to the mission of democratization

XXX Screenshot

This is another great step in our [partnership](https://huggingface.co/blog/intel), and great news for our user community, as we can now all enjoy the speedup delivered by the [Intel Xeon Ice Lake](https://www.intel.com/content/www/us/en/products/docs/processors/xeon/3rd-gen-xeon-scalable-processors-brief.html) architecture. And yes, the feature is still free!
 

> ## Inference API

Reviewer: ```suggestion
## Free Inference with Inference API
```

Reviewer: that may be a mouthful, "Free Inference API" probably better

Reviewer 2: how about "Evaluate with Inference API" ?

The [Inference API](https://huggingface.co/docs/api-inference/) is equally developer-friendly. With a simple HTTP request, you can load any hub model and predict your data with it in seconds. The model URL and a valid hub token are all you need, for example: 

```
curl https://api-inference.huggingface.co/models/xlm-roberta-base \
	-X POST \
	-d '{"inputs": "The answer to the universe is <mask>."}' \
	-H "Authorization: Bearer HF_TOKEN"
```

The Inference API is the simplest way to build a prediction API that you can immediately call from your application during development and tests. No need for a bespoke prediction service, or for a model server. In addition, you can instantly switch from one model to the next and compare their performance in your application. And guess what? The Inference API is free to use.

> ## Inference Endpoints

Reviewer: ```suggestion
## Production with Inference Endpoints
```

Once you're happy with the performance of your ML model, it's time to deploy it for production. Unfortunately, when you leave the sandbox, everything becomes a concern: security, scaling, monitoring, etc. This is where a lot of ML stumble and sometimes fall.

This is exactly why we built [Inference Endpoints](https://huggingface.co/inference-endpoints). In just a few clicks, you deploy any hub model on secure and scalable infrastructure hosted in your AWS or Azure region of choice. Additional settings include CPU and GPU hosting, built-in auto-scaling, and more.

XXX screenshot

Inference Endpoints support three security levels:

* Public: the endpoint runs in a public Hugging Face subnet, and anyone on the Internet can access it without any authentication.

* Protected: the endpoint runs in a public Hugging Face subnet, and anyone on the Internet with the appropriate organization token can access it.

* Private: the endpoint runs in a private Hugging Face subnet. It's not accessible on the Internet. It's only available through a private connection in your AWS or Azure account. This will satisfy the strictest compliance requirements.

> 

Reviewer: I would add a mention of the pricing to further differentiate with Inference API, e.g.

Inference Endpoints deploy your models onto dedicated infrastructure just for you, and your account will be billed for the time your endpoint will be up and running, depending on which compute option you choose - see [pricing](https://huggingface.co/docs/inference-endpoints/pricing) for details and billing examples.

To learn more, please read this [tutorial](https://huggingface.co/blog/inference-endpoints) and the [documentation](https://huggingface.co/docs/inference-endpoints/).

## Getting started

> It couldn't be simpler. Just log in to the Hugging Face [hub](https://huggingface.co/) and browse our [models](https://huggingface.co/models). Once you've found one that you like, you can try the Inference Widget directly on the page. Clicking on the "Deploy" button, you'll get auto-generated code to deploy the model on the Inference API, and a direct link to deploy it with Inference Endpoints.

Reviewer 2: ```suggestion
It couldn't be simpler. Just log in to the Hugging Face [hub](https://huggingface.co/) and browse our [models](https://huggingface.co/models). Once you've found one that you like, you can try the Inference Widget directly on the page. Clicking on the "Deploy" button, you'll get auto-generated code to deploy the model on the free Inference API for evaluation, and a direct link to deploy it to production with Inference Endpoints.
```

---
title: "An overview of inference solutions on Hugging Face"
thumbnail: /blog/assets/116_inference_update/widget.png
---

<h1>An Overview of Inference Solutions on Hugging Face</h1>

<div class="blog-metadata">
    <small>Published November 21, 2022.</small>
    <a target="_blank" class="btn no-underline text-sm mb-5 font-sans" href="https://github.com/huggingface/blog/blob/main/openvino.md">
        Update on GitHub
    </a>
</div>

<div class="author-card">
        <a href="https://twitter.com/julsimon">
        <img class="avatar avatar-user" src="https://aeiljuispo.cloudimg.io/v7/https://s3.amazonaws.com/moonup/production/uploads/1633343465505-noauth.jpeg?w=128&h=128&f=face" title="Julien Simon">
        <div class="bfc">
            <code>juliensimon</code>
            <span class=fullname">Julien Simon</span>
        </div>
    </a>
</div>

> Transformer models are quickly becoming the *de facto* architecture for a wide range of machine learning (ML) applications, including natural language processing, computer vision, speech, and more. Every day, developers and organizations are adopting these models to turn ideas into proof-of-concept demos, and demos into production-grade applications. 

Reviewer 3: as discussed before we don't want to emphasize just transformers any more

i'll propose a rewording in another PR

Now, let's review your inference options with Hugging Face.

## Free Inference Widget

One of my favorite features on the Hugging Face hub is the Inference [Widget](https://huggingface.co/docs/hub/models-widgets). Located on the model page, the Inference Widget lets you upload sample data and predict it in a single click. 

Here's a sentence similarity example with the `sentence-transformers/all-MiniLM-L6-v2` [model](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2):

<kbd>
  <img src="assets/116_inference_update/widget.png">
</kbd>

> 

Reviewer 3: add diffusers example as they have a ton of traction these days
